Mathlib/Lean/Expr/Basic.lean

```diff
@@ -83,6 +83,16 @@ def isPrefixOf? (pre nm : Name) : Option Name :=
   | num p' a => (isPrefixOf? pre p').map (·.num a)
   | str p' s => (isPrefixOf? pre p').map (·.str s)
 
+/--
+Declarations that are internal or automatically generated in certain ways.
+Note: this declaration also occurs as `shouldIgnore` in the Lean 4 file `test/lean/run/printDecls`.
+-/
+def isInternal' (declName : Name) : Bool :=
+  declName.isInternal ||
+  match declName with
+  | .str _ s => "match_".isPrefixOf s || "proof_".isPrefixOf s || "eq_".isPrefixOf s
+  | _        => true
```

Comment:
Could you evaluate how this compares to this?
```lean
def shouldIgnoreDecl (declName : Name) : MetaM Bool := do
  match ← findDeclarationRanges? declName with
  | some _ =>
    let env ← getEnv
    pure (declName.isInternal)
    <||> (pure <| isAuxRecursor env declName)
    <||> (pure <| isNoConfusion env declName)
    <||> isRec declName
    <||> Lean.Meta.isMatcher declName
  | none => return true
```
This particular definition comes in part from the doc-gen, but I haven't really looked too carefully into exactly what it's doing.

At the least, I think the docstring should be clear that it's guessing based on the structure of the name.

Reply:
I have to study exactly what declarations are found by `findDeclarationRanges`. It seems that the following already finds (almost?) all internal declarations, but it might also find a few more declarations that are automatically generated and weren't tagged with declaration ranges.
I'm going not change the implementation of `isInternal'` in this PR, but I agree we should consider changing it.
```lean
def isInternal'' (decl : Name) : CoreM Bool :=
  return (← findDeclarationRangesCore? decl) |>.isNone
```